File: src/main/java/me/st28/flexseries/flexlib/utils/SchedulerUtils.java
/**
 * Copyright 2016 Stealth2800 <http://stealthyone.com/>
 * Copyright 2016 Contributors <https://github.com/FlexSeries>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.st28.flexseries.flexlib.utils;

import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;

/**
 * Utility functions for the Bukkit scheduler.
 */
public final class SchedulerUtils {

    private SchedulerUtils() { }

    public static void runSynchronously(JavaPlugin plugin, Runnable runnable) {
        if (Bukkit.isPrimaryThread()) {
            runnable.run();
        } else {
            Bukkit.getScheduler().runTask(plugin, runnable);
        }
    }

    public static void runAsynchronously(JavaPlugin plugin, Runnable runnable) {
        if (Bukkit.isPrimaryThread()) {
            Bukkit.getScheduler().runTaskAsynchronously(plugin, runnable);
        } else {
            runnable.run();
        }
    }

    public static void runAsap(JavaPlugin plugin, Runnable runnable, boolean async) {
        if (async) {
            runAsynchronously(plugin, runnable);
        } else {
            runSynchronously(plugin, runnable);
        }
    }

}

File: src/main/kotlin/me/st28/flexseries/flexlib/util/SchedulerUtils.kt
/**
 * Copyright 2016 Stealth2800 <http://stealthyone.com/>
 * Copyright 2016 Contributors <https://github.com/FlexSeries>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.st28.flexseries.flexlib.util

import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin

object SchedulerUtils {

    fun runSync(plugin: JavaPlugin, function: () -> Unit) {
        if (Bukkit.isPrimaryThread()) {
            function.invoke()
        } else {
            Bukkit.getScheduler().runTask(plugin, function)
        }
    }

    fun runAsync(plugin: JavaPlugin, function: () -> Unit) {
        if (Bukkit.isPrimaryThread()) {
            Bukkit.getScheduler().runTaskAsynchronously(plugin, function)
        } else {
            function.invoke()
        }
    }

    fun runAsap(plugin: JavaPlugin, function: () -> Unit, async: Boolean) {
        if (async) {
            runAsync(plugin, function)
        } else {
            runSync(plugin, function)
        }
    }

}